### DISCUSSION TITLE
Don't call the root loader on /healthcheck route

### DISCUSSION THREAD

**Discussion by lukejagodzinski** [remix-run/remix#7543] [Q&A] [answered]
I'm fighting with this problem for a very long time already. So far, I always was able to somehow find a workaround but this time I can't. Fly.io calls `/healthcheck` route periodically and ideally it should always return `200`. However, in my app I need to always throw response with status 503 when the maintenance happens. When I `throw Response(null, {status: 503});` in the root `loader` then it will also throw 503 in the `/healthcheck`. Is it possible to prevent calling root in some routes? I'm using Remix server and not express. I guess in express I could create separate route just to handle health check.

**Comment by kiliman** [comment]
In your root loader, you can check the pathname. That way, you don't trigger the 503 response.
```ts
export function loader({ request }: DataFunctionArgs) {
  const url = new URL(request.url);
  if (url.pathname === "/healthcheck") {
    return json({ message: "ok" });
  }
  return json({ message: "hello world" });
}
```

**Reply by lukejagodzinski** [reply]
It doesn't work when I do it this way. So in the configuration where I have just `root.tsx` and `healthcheck.tsx` and when I console log in each file:
```tsx
console.log("routes/root.tsx loader()", new URL(request.url).pathname);
```

```tsx
console.log("routes/healthcheck.tsx loader()", new URL(request.url).pathname);
```

I get the following logs:
```
routes/healthcheck.tsx loader() /healthcheck
routes/root.tsx loader() /
```

So root never gets the `/healthcheck` pathname. It always get `/` which makes sense. Remix is making actually multiple requests to multiple routes to get data needed to render the page and `root` always receives `/`.

In my case, it's even more complicated because I also have `routes/_index.tsx` file that supposed to redirect user when he is not logged in. So when I bring this file back, I get different logs:

```
routes/healthcheck.tsx loader() /healthcheck
routes/_index.tsx loader() /
```

So basically it's not possible to stop work in other routes for being done. If any of the routes return non 200 code then it takes precedence (to my understanding). If I redirect in one of the parent routes, then the child route will be ignored. If I return 4XX in any of the routes it will also be used instead of the child route 200.

**Reply by lukejagodzinski** [reply]
What's even weirder is that when I have redirects in `routes/_index` it can get even different route but never `/healthcheck`

```
routes/healthcheck.tsx loader() /healthcheck
routes/_index.tsx loader() /
HEAD / 302 - - 23.512 ms
HEAD /admin 302 - - 19.497 ms
routes/root.tsx loader() /admin/users
HEAD /admin/users 200 - - 52.985 ms
GET /healthcheck 200 - - 142.402 ms
```

**Reply by lukejagodzinski** [reply]
Hmmm I've noticed one weird thing. When I call the `/healthcheck` route then all the parent routes are requested using the HEAD method. But for regular routes it's using the `GET` method. I used that and before the redirect logic I put:
```tsx
if (request.method === "HEAD") {
  throw new Response();
}
```
And it looks like it's working. I'm not sure if it's the right way to solve this problem. What I don't like about Remix is that there loader handles multiple methods. I would prefer to have one method for each method: GET, POST, PUT, HEAD, DELETE etc.

**Reply by lukejagodzinski** [reply]
OMG... I've just checked the code in the `healthcheck.tsx` file and it's making the HEAD request to home page... When I removed this call it doesn't even get the root route... I don't know why I missed that detail for so long 😀

**Reply by lukejagodzinski** [reply, marked as answer]
Actually, the final solution that I used was to modify only the `/healthcheck` route.
Leaving it here for posterity:
```tsx
// learn more: https://fly.io/docs/reference/configuration/#services-http_checks
import { type LoaderArgs } from "@remix-run/node";
import { prisma } from "~/db.server";

export async function loader({ request }: LoaderArgs) {
  const host =
    request.headers.get("X-Forwarded-Host") ?? request.headers.get("host");
  try {
    const url = new URL("/", `http://${host}`);
    // If we can connect to the database and make a simple query and make a HEAD
    // request to ourselves, then we're good.
    await Promise.all([
      prisma.user.count(),
      fetch(url.toString(), { headers: request.headers, method: "HEAD" }).then(
        function (response) {
          // Accept 503 responses as a valid response for healthcheck as server
          // is in the maintenance mode.
          if (response.status === 503) {
            return;
          }
          if (!response.ok) {
            return Promise.reject(response);
          }
          return;
        },
      ),
    ]);
    return new Response("OK");
  } catch (error: unknown) {
    console.log("healthcheck ❌", { error });
    return new Response("ERROR", { status: 500 });
  }
}
```

**Reply by kiliman** [reply]
I'm glad you figured it out. But to clarify, when Remix calls multiple loaders in parallel, there is still a _single_ request pathname. For the initial document request, a single request is passed to all the loaders. For client navigation/revalidation, there are multiple requests, but they all have the `same` request pathname (the loader is identified by the `_data` search param).

As you noticed, the reason your root was showing both a `/healthcheck` and `/` request was that the healthcheck route was calling your root route directly.

**Reply by lukejagodzinski** [reply]
Yep, thanks for the clarification!

**Comment by gaurish-saini** [comment]
Glad to find this discussion while working on a Remix v1 legacy codebase deployed on fly.io. This was really helpful to understand things better, even after 2 years!
Thank you :)
